### PR TITLE
Proposal: Provide a Common MQTT Mapper

### LINKED ISSUE #5588: [OSPP] Provide A Common MQTT Mapper
**What would you like to be added/modified**:

- [ ] Provide a proposal;
- [ ] Craete a MQTT Mapper project and is working properly;
- [ ] The attribute value in MQTT message can be parsed by the CRDs of DeviceModel and Device, these serialization methods need to be completed: json, yaml and xml;

**Why is this needed**:

The MQTT protocol is the common device protocol in most device hardware markets. We want to have a common MQTT Mapper to handle these devices using the MQTT protocol, and it can support getting attribute values from the common serialization methods(i.e. json, yaml and xml). In order to fulfill the above requirements, we need you to complete these tasks:
1. We expect you to use the latest mapper-framework to add a Mapper for MQTT protocol to the project kubeedge/mappers-go;
2. We expect this Mapper can read the message from the topic of the specified MQTT service, And the attribute value can be parsed from the message by common serialization methods (i.e. json and xml);
3. (Optional)We expect you to provide a more scalable solution to parse attribute values from more custom messages;

**Refer**:

- https://github.com/kubeedge/kubeedge/blob/master/docs/proposals/device-crd-v1beta1.md
- https://github.com/kubeedge/kubeedge/tree/master/staging/src/github.com/kubeedge/mapper-framework

### PR DESCRIPTION
**Pull request by fuchendou**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #5588

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
+---
+title: Mapper Provide A Common MQTT Mapper
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:2]: This description is a bit strange, how about using Provide A Common MQTT Mapper directly?

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+- Developers: Developers are needed to implement and test the MQTT Mapper.
+
+- IoT Device Manufacturers and Operators: IoT device manufacturers and operators can use this universal MQTT Mapper to manage device communications.
+
+### Design Details
+
+##### Architecture and Modules
+
+        The architecture for Mapper Provide a Common Mapper will be modular to ensure scalability and maintainability. The main components include:
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:53]:
```suggestion
        The architecture of providing a common MQTT mapper will be modular to ensure scalability and maintainability. The main components include:
```

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        The architectures and related concepts are shown in the below figure. In the figure, we assume that in the cloud-side-end scenario, device state modifications can be initiated by the cloud.
+
+<img title="相对路径演示" src="../images/proposals/mapper-provide-a-common-MQTT-mapper-architecture.jpg" alt="本地路径" width="546" data-align="center">
+
+##### Base Environment Setup
+
+1. Generate Mapper files through the Mapper-Framework framework
+   
+   1)Download kubeedge-master.zip from the official website, take the mapper-framewokr under the staging folder in it, and transfer it to the VM master node.
+
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:68]: The mapper-framework located in the staging directory of the kubeedge main repo is in the development process. You can download the stable release of mapper-framework at https://github.com/kubeedge/mapper-framework.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+   spec:
+     properties:
+       - name: temp
+         description: beta1-model
+         type: INT
+         accessMode: ReadWrite
+         maximum: "100"
+         minimum: "1"
+         unit: "Celsius"
+     protocol: modbus
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:97]:
```suggestion
     protocol: mqtt
```

This proposal is to provide a mapper for the mqtt protocol, so the protocol name here should be mqtt

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+             offset: 2
+             limit: 1
+             scale: 1
+             isSwap: true
+             isRegisterSwap: true
+     protocol:
+       protocolName: modbus
+       configData:
+         ip: 172.17.0.3
+         port: 1502
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:148]: ditto

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                 fieldKey: beta1test
+         visitors:
+           protocolName: modbus
+           configData:
+             register: "HoldingRegister"
+             offset: 2
+             limit: 1
+             scale: 1
+             isSwap: true
+             isRegisterSwap: true
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:143]: Here are fields related to the modbus protocol, which should be replaced by fields common to the mqtt protocol.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+        The architecture for Mapper Provide a Common Mapper will be modular to ensure scalability and maintainability. The main components include:
+
+- Base Environment Setup: Utilizes the latest mapper-framework to create a MQTT mapper for standardized operatons and interactions.
+
+- Main algorithm design: Parses the attribute values from the message using common serialization methods(JSON, YAML, XML).
+
+        The architectures and related concepts are shown in the below figure. In the figure, we assume that in the cloud-side-end scenario, device state modifications can be initiated by the cloud.
+
+<img title="相对路径演示" src="../images/proposals/mapper-provide-a-common-MQTT-mapper-architecture.jpg" alt="本地路径" width="546" data-align="center">
```
Comment (wbc6080) [inline docs/proposals/mapper- provide-a- common- MQTT- mapper.md:61]: The communication method between devicetwin and mapper in the architecture diagram is wrong. Now it relies on dmi in devicetwin and defines the grpc interface for implementation, rather than mqtt broker.

edge

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 2 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md:53]: Thank you for pointing out the unclear design ideas and related errors in the proposal, I will promptly revise my proposal in response to your comments.

**Comment by magichan** [comment]
@fuchendou  you may need assign the releated issue #5588  to youself. 

That will help anyone who care about project activity.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 1 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md:2]: Ok, I will take you advice.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 3 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md:68]: Ok, I understand what you mean, I will modify the expression here, in practice we use Kubeedge v1.17.0 version as the base environment for generating mapper.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 4 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md]: Ok, I'll all change the model and instance files based on your comments.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 5 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md]: Ok, I'll all change the model and instance files based on your comments.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 6 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md]: Ok, I'll all change the model and instance files based on your comments.

File: docs/proposals/mapper- provide-a- common- MQTT- mapper.md (hunk 7 above)
Comment (fuchendou) [reply docs/proposals/mapper- provide-a- common- MQTT- mapper.md:61]: Ok, I will use the architecture diagram you provided me as an overall description of Kubeedge.

File: docs/proposals/Provide-a- Common- MQTT- mapper.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+##### Base Environment Setup
+
+1. Generate Mapper files through the Mapper-Framework framework
+   
+   1)Download kubeedge-master.zip from the official website, take the mapper-framewokr under the staging folder in it, and transfer it to the VM master node.
+   
+   2)Execute a command: 
+   
+   Here we use Kubeedge v1.17.0 as the base environment to generate the mapper, or you can download the latest version of the mapper directly from the repository ( [<u>https://github.com/kubeedge/mapper-framework)</u>](https://github.com/kubeedge/mapper-framework).
```
Comment (wbc6080) [inline docs/proposals/Provide-a- Common- MQTT- mapper.md:71]: There is some confusion in the expression here. There is no need to download kubeedge-master.zip, just clone the mapper-framework repo.

File: .gitignore [hunk 9]
```diff
@@ -18,3 +18,12 @@ keadm/keadm
...
 build/edge/tmp
+.lightly
+.c_bin
+.classes
+logs
+*.log
+node_modules
+dist
+dist-ssr
+*.local
```
[the remark this answers is not in the document]
Comment (fuchendou) [reply .gitignore:29]: Ok, I will delete them.

File: docs/proposals/Provide-a- Common- MQTT- mapper.md (hunk 8 above)
Comment (fuchendou) [reply docs/proposals/Provide-a- Common- MQTT- mapper.md:71]: Sorry, I should have used “clone” instead of “download”.

File: .gitignore [hunk 10]
```diff
@@ -17,4 +17,4 @@ edgesite/edgesite
 keadm/keadm
 _output
 _tmp
-build/edge/tmp
+build/edge/tmp
```
Comment (fujitatomoya) [inline .gitignore:20]: looks like unnecessary change here? or git does a funny thing?

**Comment by fujitatomoya** [comment]
issue title has a typo,

> Add a Proposal about Privide a Common MQTT Mapper

should be, `Add a Proposal about Provide a Common MQTT Mapper`

File: docs/proposals/Provide-a- Common- MQTT- mapper.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+   
+   Here we use Kubeedge v1.17.0 as the base environment to generate the mapper, or you can clone the latest version of the mapper directly from the repository ( [<u>https://github.com/kubeedge/mapper-framework)</u>](https://github.com/kubeedge/mapper-framework).
+   
+   ```go
+   make generate
+   ```
+   
+   3)Enter the project name of the Mapper: mqtt
+   
+   <img title="" src="../images/proposals/Provide-a-common-MQTT-mapper-pic.jpg" alt="" data-align="center">
```
Comment (fujitatomoya) [inline docs/proposals/Provide-a- Common- MQTT- mapper.md:79]: Does this have to be a image? i would use text instead, would be easier to change and track the history.

**Review by fujitatomoya — COMMENTED** [review]
a couple of minor comments.

**Comment by fuchendou** [comment]
Thanks for the tip!

File: .gitignore (hunk 10 above)
Comment (fuchendou) [reply .gitignore:20]: I think it's auto-generated and I didn't make any changes to it. Maybe it's because I introduced unnecessary configurations when I clone the project with the cloud code platform before and re-clone it locally. But it was not an intentional event.

**Comment by WillardHu** [comment]
Rebase the latest master
```bash
git fetch <kubeedge remote>
git rebase <kubeedge remote>/master
```

**Review by wbc6080 — COMMENTED** [review]
/lgtm
cc @WillardHu @fisherxu @Shelley-BaoYue

**Review by Shelley-BaoYue — COMMENTED** [review]
/lgtm

**Review by fisherxu — COMMENTED** [review]
/approve
